server/src/main/java/io/crate/expression/operator/any/AnyNeqOperator.java

```diff
@@ -65,7 +66,21 @@ protected Query refMatchesAnyArrayLiteral(Function any, Reference probe, Literal
 
         BooleanQuery.Builder andBuilder = new BooleanQuery.Builder();
         for (Object value : (Iterable<?>) candidates.value()) {
-            andBuilder.add(EqOperator.fromPrimitive(probe.valueType(), probe.column().fqn(), value), BooleanClause.Occur.MUST);
+            var fromPrimitive = EqOperator.fromPrimitive(
+                probe.valueType(),
+                probe.column().fqn(),
+                value,
+                // IntField.newExactQuery (query chosen when indexed and doc-values available) causes TooManyNestedClauses exception.
+                // For example, 8192 (default for indices.query.bool.max_clause_count) IntField.newExactQuery triggers the exception
+                // unlike other queries like SortedNumericDocValuesField.newSlowExactQuery, IntPoint.newExactQuery, or a generic query.
+                // Since refMatchesAnyArrayLiteral retrieves only a single ref so any query can be used.
+                // But assuming for the worst, generic query should be used hence no doc-values/indexes.
+                false,
```

Comment:
we need to use doc values and index if available, there is the restriction of boolean queries but you can override it, see:
https://github.com/crate/crate/blob/master/server/src/test/java/io/crate/integrationtests/LuceneQueryBuilderIntegrationTest.java#L51 and https://github.com/crate/crate/blob/master/server/src/test/java/io/crate/integrationtests/LuceneQueryBuilderIntegrationTest.java#L308

Reply:
Thank you, I added this workaround due to the test case you pointed to. When the limit is set to `10_000` we would want to be able to compare a ref to an array literal with `10_000` elements. However, if we use `IntField.newExactQuery` we end up with about 3 times more queries. Part of the cause seems to be `IndexOrDocValuesQuery.visit` calling both `indexQuery.visit` and `dvQuery.visit` where each visit is counted as `1`.
I agree that we should use doc values and indexes, I will look further.